gc/base/ParallelDispatcher.cpp

```diff
@@ -661,9 +660,18 @@ MM_ParallelDispatcher::contractThreadPool(MM_EnvironmentBase *env, uintptr_t new
 	if (newThreadCount < _threadCountMaximum) {
 		Trc_MM_ParallelDispatcher_contractThreadPool_Attempt();
 
+		_inShutdown = true;
```

Comment:
Will moving this assignment outside the monitor's scope cause any data consistency issue?

Reply:
It's a valid concern, but to my understanding there is little (or no) relationship between this flag and  _workerThreadMutex down below (which deals mostly with (de)activating a worker thread within a parallel GC task). 

This flag is more associated with _dispatcherMonitor (which deals with starting/stoping a physical worker thread). Probably to detect a rare scenario when shutdown is requested early and GC threads are late starting up, so that the 2 processes overlap. 

Indeed, in a real shutdown (in shutDownThreads()) we do notify on that _dispatcherMonitor just after setting this flag to true. Even there setting the flag is not done  under the notify critical section, but just ahead of it, what is probably not correct to start with.

So, if this is to be within a  critical section, it should be under:
```
	omrthread_monitor_enter(_dispatcherMonitor);
	_inShutdown = true;
	omrthread_monitor_notify_all(_dispatcherMonitor);
	omrthread_monitor_exit(_dispatcherMonitor);
```

But that would deal with a completely different (potential) problem, not the one this one PR deals with. So, I believe moving it up (outside of _workerThreadMutex)  does not really make it any better or worse.

But let's see what @RSalman thinks about it before we decide if we want to do anything in this PR.

For future, one thing we can try is to artificially delay initial creation of GC threads to see if it can cause problems with thread shutdown, or thread pool contraction.

Reply:
@amicic is correct. There is no concern about the ordering here, it's always been done in this order during a real shutdown (during VM shutdown, in shutDownThreads()).

Reply:
Just to expand on this point:

>It's a valid concern, but to my understanding there is little (or no) relationship between this flag and _workerThreadMutex down below (which deals mostly with (de)activating a worker thread within a parallel GC task).

There is somewhat of a relationship between the two. When we enter the mutex, we only do a notify_all only when the flag is set. Seen here in `cleanupAfterTask`:

```C++
void
MM_ParallelDispatcher::cleanupAfterTask(MM_EnvironmentBase *env)
{
	omrthread_monitor_enter(_workerThreadMutex);
	
	_workerThreadsReservedForGC = false;
	Assert_MM_true(_threadsToReserve == 0);
	_task = NULL;
	
	if (_inShutdown) {
		omrthread_monitor_notify_all(_workerThreadMutex);
	}
	
	omrthread_monitor_exit(_workerThreadMutex);
}
```

Reply:
Regardless, the order is not an issue here